### PR TITLE
`Match.groups()` return sequence can contain `None` values

### PR DESCRIPTION
**Pull request by hukkin**
Simple example regex to produce `None` in `Match.groups()` output can be `foo(bar)?`. The capturing group is optional so `groups()` can have a `None`.

### CODE REVIEW

**Comment by Akuli** [comment]
As you can see from `mypy_primer` output, this breaks code that always gets strings from `.groups()`.

Consider using `Union[Any, AnyStr]`, so that it will be actually `Union[Any, str]`. It looks weird, but it is a way to say "this can be treated as `str` without causing type errors, but it could be anything else too". See also #5269.

**Comment by Akuli** [comment]
Just found #3902. It's about `.group()`, but seems to apply here too.

File: stdlib/typing.pyi [hunk 1]
```diff
@@ -568,7 +568,7 @@ class Match(Generic[AnyStr]):
     def group(self, __group: Union[str, int] = ...) -> AnyStr: ...
     @overload
     def group(self, __group1: Union[str, int], __group2: Union[str, int], *groups: Union[str, int]) -> Tuple[AnyStr, ...]: ...
-    def groups(self, default: AnyStr = ...) -> Sequence[AnyStr]: ...
+    def groups(self, default: AnyStr = ...) -> Sequence[Optional[AnyStr]]: ...
```
Comment (hukkinj1) [inline stdlib/typing.pyi:571]:
```suggestion
    def groups(self, default: AnyStr = ...) -> Sequence[Union[Any, AnyStr]]: ...
```

**Comment by Akuli** [comment]
LGTM, but I'd like to get the opinion of another maintainer too.

**Comment by hukkinj1** [comment]
Changed to `Union[Any, AnyStr]`, although my understanding is that this isn't any different from simply `Any`?

Anyways, at least I'd prefer false positives over false negatives, so I'd be happy with `Any`.

**Comment by Akuli** [comment]
There is a subtle difference: if you do something with `Union[Any, str]`, it must be possible to do that with `str`. For example, `foo + 123` succeeds when `foo` has type `Any`, but fails when `foo` has type `Union[Any, str]`, because `str + int` doesn't make sense.

**Comment by hukkinj1** [comment]
:+1:  Thanks for the clafification and quick review!

**Comment by Akuli** [comment]
As a side note, it seems like `Tuple[blah, ...]` would be a more appropriate return type than `Sequence[blah]`, since it is always a tuple (unless I don't see some gotcha here). But that can be a separate PR.

**Comment by Akuli** [comment]
I looked through the new mypy primer errors. I was too lazy to write notes about every repo I checked, but none of the errors look too bad to me.

**Comment by srittau** [comment]
I am really unsure about this. On the one hand I am unhappy that we can't use a better return value, ideally optional, on the other hand I don't think the following (reduced) code from sphinx should be a type error:

```python
        matched = py_ext_sig_re.match(self.name)
        explicit_modname, path, base, args, retann = matched.groups()
        parents = path.rstrip('.').split('.') if path else []
```

From the first primer hit above.

**Comment by hukkinj1** [comment]
Isn't the `"Need type annotation"` type of "soft" error (the error you pointed out in Sphinx) better than mypy insisting that the user is wrong when handling a `None`?

How about a return type `Any` then, if there's no perfect way around the false negatives?

**Comment by srittau** [comment]
I am not a fan of either the status quo or the change. This means I am fine with this change, but I would also be fine with leaving it as is.

**Comment by srittau** [comment]
Thank you for the PR. This has been superseded by #5557, which incorporated the ideas from this PR.

**Pull request closed without merging**